external-crates/move/crates/move-analyzer/prettier-plugin/src/printer.ts

```diff
@@ -53,6 +54,81 @@ export function print(path: AstPath, options: ParserOptions, print: printFn) {
                 indent([line, path.call(print, 'namedChildren', 2)]),
                 ';',
             ]);
+        case 'struct_definition':
+            // type parameters are on separate lines if they don't fit on one, but fields are always
+            // on separate lines:
+            //
+            // struct SomeStruct<T1: key, T2: drop> has key {
+            //     f: u64,
+            // }
+            //
+            // struct AnotherStruct<
+            //     T1: store + drop + key,
+            //     T2: store + drop + key,
+            //     T3: store + drop + key,
+            // > has key, store {
+            //     f1: u64,
+            //     f2: u64,
+            // }
+            return [
+                'struct ',
+                path.call(print, 'namedChildren', 0),
+                path.call(print, 'namedChildren', 1),
+                path.call(print, 'namedChildren', 2),
+                path.call(print, 'namedChildren', 3),
+            ]
+        case 'native_struct_definition':
+            // same formatting as "regular" struct but (of course) without fields
+            return [
+                'struct ',
+                path.call(print, 'namedChildren', 0),
+                path.call(print, 'namedChildren', 1),
+                path.call(print, 'namedChildren', 2),
+                ';',
+            ]
+        case 'ability_decls':
+            return [
+                ' has ',
+                join(', ', path.map(print, 'namedChildren'))
+            ];
+        case 'type_parameters':
+            const tparams = Symbol('tparams');
+            return [
+                '<',
+                group([
+                    indent(softline),
+                    indent(join([',', line], path.map(print, 'namedChildren'))),
+                ], {id: tparams}),
+                ifBreak([',', softline], '', {groupId: tparams}),
+                '>',
+            ];
+        case 'type_parameter':
+            let abilities = [];
+            for (let i = 1; i < node.namedChildren.length; i++) {
+                abilities.push(path.call(print, 'namedChildren', i));
+            }
```

Comment:
We start at `i = 1` because this targets abilities in some syntax like `T1: key` and `i = 0` refers to `T1` right?

Reply:
Exactly right! Do, node.namedChildren[0] is always there, the rest is optional. Which had me realize that there is a bug here actually - if there are no abilities, `:` will still be printed (fixed and added a test for it).